Cron/KlSyncs.php

```diff
@@ -0,0 +1,189 @@
+     */
+    public function clean()
+    {
+        $syncCollection = $this->_syncCollectionFactory->create();
+        $idsToDelete = $syncCollection->getIdsToDelete('SYNCED');
+
+        $syncCollection->deleteRows($idsToDelete);
+
+        $klSyncTableSize = $syncCollection->count();
+        if ($klSyncTableSize > 10000 ){
+            $this->_klaviyoLogger->log("Klaviyo Clean Sync: kl_sync table size greater than 10000, currently sitting at $klSyncTableSize");
+        }
+
+        return;
+    }
+
+    /**
+     * Retry Cron job retries all rows marked for retry in kl_sync table
+     * @throws Exception
+     */
+    public function retry()
+    {
+        $syncCollection = $this->_syncCollectionFactory->create();
+        $groupedRows = $this->getGroupedRows( $syncCollection->getRowsForSync('RETRY')->getData() );
+
+        $this->sendUpdatesToApp($groupedRows, true);
+
+        return;
+    }
+
+    /**
+     * Sends Webhook or Track API requests based on the topic of each row
+     * and creates a response manifest to updates row status
+     * @param $groupedRows
+     * @param bool $isRetry
+     * @throws Exception
+     */
+    private function sendUpdatesToApp($groupedRows, bool $isRetry = false)
+    {
+        $webhookTopics = ['product/save']; //List of topics that use webhooks
+        $trackApiTopics = ['Added To Cart']; //List of topics that use the Track API
+
+        $responseManifest = ['1' => [], '0' => []];
+
+        foreach($groupedRows as $topic => $rows){
+            if (in_array($topic, $webhookTopics) && !empty($rows)) {
+                foreach($rows as $row) {
+                    $response = $this->_webhookHelper->makeWebhookRequest(
+                        $row['topic'],
+                        $row['payload'],
+                        $row['klaviyo_id']
+                    );
+                    if (!$response) {$response = '0';}
+
+                    array_push( $responseManifest["$response"], $row['id']);
+                }
+            }
+
+            if (in_array($topic, $trackApiTopics) && !empty($rows)) {
+                foreach($rows as $row) {
+                    $response = $this->_dataHelper->klaviyoTrackEvent(
+                        $row['topic'],
+                        json_decode($row['user_properties'], true ),
+                        json_decode($row['payload'], true ),
+                        strtotime($row['created_at'])
+                    );
+                    if (!$response) {$response = '0';}
+
+                    array_push( $responseManifest["$response"], $row['id']);
+                }
+            }
+        }
+        $this->updateRowStatuses( $responseManifest, $isRetry );
+    }
+
+    /**
+     * Update statues of rows to SYNCED, RETRY and FAILED based on response and if Retry cron run
+     * @param $responseManifest
+     * @param $isRetry
+     */
+    private function updateRowStatuses( $responseManifest, $isRetry )
```

Comment:
whats the idea behind the statuses.  If it fails the first time you assign it 'RETRY' and if second time 'FAILED'?

Reply:
that's correct